theano/sandbox/cuda/var.py

```diff
@@ -211,3 +213,40 @@ def float32_shared_constructor(value, name=None, strict=False,
     rval.get_value_return_ndarray = get_value_return_ndarray
 
     return rval
+
+
+def CudaNdarraySharedVariable_unpickler(*npa):
+    if (config.experimental.unpickle_shared_gpu_on_cpu and
+        config.device == 'cpu'):
+        # directly return numpy array
+        warnings.warn(
+            "config.experimental.unpickle_shared_gpu_on_cpu is set to True."
+            " Unpickling CudaNdarraySharedVariable as TensorSharedVariable."
+        )
+        cls = theano.tensor.sharedvar.TensorSharedVariable
+        type = theano.tensor.TensorType(
+            dtype=npa[1].dtype, broadcastable=npa[1].broadcastable)
+        return cls(name=npa[0],
+                   type=type,
+                   value=numpy.asarray(npa[2].data),
+                   strict=None)
```

Comment:
A shared variable have the data in a Container object.. Here I do not keep
the container intact. It could be shared in some cases. This is what it
refer.

On Tue, Mar 24, 2015 at 9:16 AM, Bart van Merriënboer <
notifications@github.com> wrote:

> In theano/sandbox/cuda/var.py
> <https://github.com/Theano/Theano/pull/2339#discussion_r27030978>:
>
> > +
> > +def CudaNdarraySharedVariable_unpickler(*npa):
> > +    if (config.experimental.unpickle_shared_gpu_on_cpu and
> > +        config.device == 'cpu'):
> > +        # directly return numpy array
> > +        warnings.warn(
> > +            "config.experimental.unpickle_shared_gpu_on_cpu is set to True."
> > +            " Unpickling CudaNdarraySharedVariable as TensorSharedVariable."
> > +        )
> > +        cls = theano.tensor.sharedvar.TensorSharedVariable
> > +        type = theano.tensor.TensorType(
> > +            dtype=npa[1].dtype, broadcastable=npa[1].broadcastable)
> > +        return cls(name=npa[0],
> > +                   type=type,
> > +                   value=numpy.asarray(npa[2].data),
> > +                   strict=None)
>
> I'm going to be a Theano noob here if that's okay. As I was suggesting in bartvm/blocks#474
> (comment)
> <https://github.com/bartvm/blocks/issues/474#issuecomment-85510424>, I'd
> like to intercept pickling of shared variables in Blocks and save them to
> separate NumPy files with sensible names. What does it mean exactly to
> break "aliasing of memory" and what are the consequences?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2339/files#r27030978>.
>